src/Init/Data/Cast.lean

```diff
@@ -0,0 +1,101 @@
+
+This typeclass was introduced to a uniform simp normal form for such morphisms.
+
+Without such a typeclass, we would have coercions such as `Int.ofNat`,
+but also later in Mathlib the generic coercion from `Nat` into any semiring (including `Int`),
+and we would need to use `simp` to move between them.
+However `simp` lemmas expressed using a non-normal form on the LHS would then not fire.
+
+In order to avoid this, we introduce this typeclass as early as possible,
+ensure it is a higher priority coercion that built-in coercions such as `Int.ofNat`,
+and write all of our simp lemmas in terms of it.
+
+Typically different instances of this class for the same target type `R` are definitionally equal,
+and so differences in the instance do not block `simp` or `rw`.
+
+The same discussion applied to `Int`, so we also introduce `IntCast`.
+
+## Note about coercions into arbitrary types:
+
+Coercions such as `Nat.cast` that go from a concrete structure such as
+`Nat` to an arbitrary type `R` should be set up as follows:
+```lean
+instance : CoeTail Nat R where coe := ...
+instance : CoeHTCT Nat R where coe := ...
+```
+
+It needs to be `CoeTail` instead of `Coe` because otherwise type-class
+inference would loop when constructing the transitive coercion `Nat → Nat → Nat → ...`.
+Sometimes we also need to declare the `CoeHTCT` instance
+if we need to shadow another coercion
+(e.g. `Nat.cast` should be used over `Int.ofNat`).
+
+-/
+
+/-- Type class for the canonical homomorphism `Nat → R`. -/
+class NatCast (R : Type u) where
+  /-- The canonical map `Nat → R`. -/
+  protected natCast : Nat → R
+
+instance : NatCast Nat where natCast n := n
+instance : NatCast Int where natCast n := Int.ofNat n
+
+/--
+Canonical homomorphism from `Nat` to a type `R`.
+
+It contains just the function, with no axioms.
+
+The prototypical example is `Int.ofNat`.
+
+In Mathlib there will be such a homomorphism whenever `R` is an additive monoid with a `1`.
+-/
+@[coe, reducible, match_pattern] protected def Nat.cast {R : Type u} [NatCast R] : Nat → R :=
+  NatCast.natCast
+
+-- see the notes about coercions into arbitrary types in the module doc-string
+instance [NatCast R] : CoeTail Nat R where coe := Nat.cast
+
+-- see the notes about coercions into arbitrary types in the module doc-string
+instance [NatCast R] : CoeHTCT Nat R where coe := Nat.cast
+
+/-- This instance is needed to ensure that `instCoeNatInt` from core is not used. -/
+instance : Coe Nat Int where coe := Nat.cast
+
+/-- Type class for the canonical homomorphism `Int → R`. -/
+class IntCast (R : Type u) where
+  /-- The canonical map `Int → R`. -/
+  protected intCast : Int → R
+
+instance : IntCast Int where intCast n := n
+
+/--
+Canonical homomorphism from `Nat` to a type `R`.
+
+It contains just the function, with no axioms.
+
+The prototypical example is `Int.ofNat`.
+
+In Mathlib there will be such a homomorphism whenever `R` is an additive group with a `1`.
```

Comment:
```suggestion
```
Forward references to downstream projects seems like a bad idea to me - I'd like to avoid the maintenance headache of keeping it all consistent!